Support certificate role mappings

sberyozkin opened the pull request:
Fixes #23683.

Tests to follow, also I thought I could avoid registering an augmentor and just pass the roles directly to the default X509 identity provider.

If these mappings are not comping from the file/classpath resource then a custom SecurityIdentityAugmentor will help.

I'll open for review later but early feedback is welcome

sberyozkin committed:
Support certificate role mappings

michalvavrik commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java:
```diff
@@ -76,4 +82,8 @@ public Set<Class<? extends AuthenticationRequest>> getCredentialTypes() {
         return Uni.createFrom().item(new HttpCredentialTransport(HttpCredentialTransport.Type.X509, "X509"));
     }
+
+    public void setRoleMappings(Map<String, Set<String>> roles) {
```
Please make this package friendly (it's set in same package) because it makes me really nervous anyone can set roles. Also if it were really public, it would have to consider concurrency. Thanks

sberyozkin replied:
Sure, I've just removed `public`...

>it makes me really nervous anyone can set roles

Anyone can set them on the identity in any case with augmentors :-) but sure, I agree it is better avoid `public`

michalvavrik replied:
Yeah, I don't have clear concept how access control should work, but I don't like to give away anything (here `public`) unless I have a good reason. Thanks for change.

sberyozkin commented:
I wanted to modify `integration-tests/vertx-http` but it is the only test which tests a truststore alias option while for this PR tests the whole truststore should be trusted to test the client certificates with different mappings

michalvavrik commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:
```diff
@@ -362,4 +373,47 @@ private synchronized void setPatchMatchingPolicyEnabled() {
+                log.warnf("Unable to read roles mappings from %s:%s", rolesPath, e.getMessage());
+            }
+            InstanceHandle<MtlsAuthenticationMechanism> mtls = Arc.container().instance(MtlsAuthenticationMechanism.class);
+            if (mtls.isAvailable()) {
```
this is **fine**, let's keep it, but tbh I tried to find when this could happen and I can't think of any situation unless there is bug in Quarkus

michalvavrik commented on integration-tests/mtls-certificates/pom.xml:
```diff
@@ -0,0 +1,166 @@
+    <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy</artifactId>
```
This is not important, but out of curiosity: why not RESTEasy Reactive? It's preferred REST stack.

sberyozkin replied:
@michalvavrik Copy and paste :-), first I tested it all in `integration-tests/vertx-http`, changes some of the existing tests there and then decided to copy it all into a new test module :-)

michalvavrik commented on integration-tests/mtls-certificates/pom.xml:
```diff
@@ -0,0 +1,166 @@
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-vertx-http</artifactId>
```
`io.quarkus:quarkus-vertx-http` is not necessary, as RESTEasy  won't ever work without it anyway (it's transitive dependency).

michalvavrik commented on integration-tests/mtls-certificates/src/test/java/io/quarkus/it/vertx/CertificateRoleMappingTest.java:
```diff
@@ -0,0 +1,60 @@
+import io.restassured.builder.RequestSpecBuilder;
+import io.restassured.specification.RequestSpecification;
+
+@QuarkusTest
```
Considering you are adding whole new module to integration tests (and not just junit test into extensions module) I think you should also add `@QuarkusIntegrationTest`

sberyozkin replied:
@michalvavrik Good catch, I have it, forgot to add

sberyozkin replied:
@michalvavrik I wanted to modify `integration-tests/vertx-http` but as I said it tests that only a single certificate is trusted and I needed at least two to test, which is why I thought of introducing a new integration tests module. I suppose having a native test for a feature request with Quarkus JIRA is worth it

michalvavrik replied:
Sure, makes sense.

michalvavrik reviewed (commented):
LGTM, few notes, two important, others you can ignore

michalvavrik commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java:
```diff
@@ -76,4 +82,8 @@ public Set<Class<? extends AuthenticationRequest>> getCredentialTypes() {
     }
+
+    void setRoleMappings(Map<String, Set<String>> roles) {
+        this.roles = roles;
```
and ideally so should be nested set, but don't want to be too annoying

michalvavrik replied:
let's go with immutable map

sberyozkin commented:
Thanks @michalvavrik, let me deal with the comments, all of them make sense

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:
```diff
@@ -362,4 +373,47 @@ private synchronized void setPatchMatchingPolicyEnabled() {
+                    rolesResource = rolesPath.toUri().toURL();
+                } catch (MalformedURLException e) {
+                    throw new ConfigurationException(
+                            "quarkus.http.auth.mtls-certificate-role-properties is malformed",
```
One thing I'm not sure about is whether it should be `quarkus.http.auth.certificate-role-properties` or `quarkus.http.ssl.certificate-role-properties` - I did not put it into the `.ssl.` one because I thought it was about  (M)TLS specifics. 
The other thing is that I should probably rename it to `certificate-role-mappings`

Ladicek commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java:
```diff
@@ -24,6 +25,16 @@ public class AuthRuntimeConfig {
+     * Use it only if the mTLS authentication mechanism is enabled with either
+     * `quarkus.http.ssl.client-auth=required` or `quarkus.http.ssl.client-auth=request`.
+     * <p/>
+     * Properties file is expected to have the `CN:role1,role,...,roleN` format.
```
I know the `.properties` format also supports `:` as a key/value separator, but we use `=` everywhere AFAIK, so we should do here as well I think?

```suggestion
     * Properties file is expected to have the `CN=role1,role,...,roleN` format.
```

sberyozkin replied:
@Ladicek Sure, this is also used in the tests and suggested here, https://github.com/quarkusio/quarkus/issues/23683#issuecomment-1818385631, so not sure where I found `:`, and good to know it is actually expected to work with Properties :-)

Ladicek commented:
I'm sure someone will come in the future asking to use a different certificate attribute :-), but CN seems good enough for a start.

Ladicek commented on integration-tests/mtls-certificates/src/main/resources/application.properties:
```diff
@@ -0,0 +1,10 @@
+quarkus.http.ssl.certificate.trust-store-file=server-truststore.jks
+quarkus.http.ssl.certificate.trust-store-password=password
+quarkus.http.ssl.client-auth=REQUIRED
+quarkus.http.auth.certificate-role-properties=role-mappings.txt
```
This is a good question. Woudln't we want to support both styles of configuration: embedded in main `application.properties` and in an extra `.properties` file, similarly to https://quarkus.io/guides/security-properties ?

sberyozkin replied:
@michalvavrik @Ladicek Good points, thanks, the file is supported because this is what the customer already has, so asking them to move them to the explicit declarations in application.properties would not be ideal, but yeah, worth supporting too, I'll have a look

michalvavrik replied:
@sberyozkin is there a difference for customers between `quarkus.http.auth.certificate-role-properties=role-mappings.txt` and `quarkus.config.locations=path-to-file-role-mappings.txt` ? I didn't try that as I'm busy today, but I think it is worth to consider that there can be no actual difference (please try it to determine if that is a valid point). Also if customers prefer colon format, they can use YAML extension. Thanks

sberyozkin replied:
@michalvavrik Not sure `quarkus.config.locations` works well, it is creating some kind of config indirection, IMHO having a dedicated property is not bad; `:` is fixed by Ladislav, `=` is suggested now

michalvavrik replied:
@sberyozkin hey difference would be config property names prefix. Also with `io.smallrye.config.source.file.FileSystemConfigSource` you could 
 `io.smallrye.config.source.file.FileSystemConfigSource#getValue` have a full control in dropping config prefix (`quarkus.http.auth.certificate-role-properties.` in overridden method, adding prefix to keys by overriding `getPropertyNames`) and user's would be able to leverage SR Config features. 

I'm fine if you just add support for config based roles.

michalvavrik replied:
IMO keeping what you have is the simplest.

sberyozkin replied:
@michalvavrik, I've been thinking about it, let me clarify something, so this PR is about mapping certificate CN names (which is only one, though primary certificate attribute) to roles, so with

`quarkus.http.auth.certificate-role.client=user,admin`

it would introduce authentication specific (MTLS) specific property (`certificate-role`) into a generic HTTP security policy configuration, and I'm not sure now we should do it.

The other thing is, something I've found out with #37152 (and I've now removed the related code), CN can be complex, they may have tricky characters, so ended up doing something like:
```
quarkus.oidc.certificate-chain.root-aliases.myCA=Root CA
quarkus.oidc.certificate-chain.endpoint-authorities.myCA=www.quarkustest.com
```
where I was mapping the CA name to some alias, note, I had to set `Root CA` as a map value, as there were issues either with `Root CA` or `www.quarkustest.com` (this is the subject cert CN) set as map keys at the config level.

I see, at #23683 I was also thinking about some maps at the config level, but after #37152 I decided to leave it, if necessary till later.

My proposal is this: since we agree that supporting file-based mapping is fine and meets a concrete requirement, lets have this PR be specific about that requirement only. And I can open an enhancement request to complement it with the mapping at the config level, and I'll copy your proposal there, and we'll keep it recorded, how does it sound ?

sberyozkin replied:
Hi @michalvavrik, I'll create an enhancement request once this PR gets one more approval and I can merge it

dmlloyd commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:
```diff
@@ -362,4 +373,50 @@ private synchronized void setPatchMatchingPolicyEnabled() {
+            }
+            Properties rolesProps = new Properties();
+            try (InputStream is = rolesResource.openStream()) {
+                rolesProps.load(is);
```
Whenever we read a properties file from the user, we should always wrap it with an `InputStreamReader` using `StandardCharsets.UTF_8` for the charset and load it from that using the `Properties#load(Reader)` overload.

It might also be prudent to add a `BufferedReader` and/or `BufferedInputStream` layer.

dmlloyd commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java:
```diff
@@ -24,6 +25,16 @@ public class AuthRuntimeConfig {
+     * Use it only if the mTLS authentication mechanism is enabled with either
+     * `quarkus.http.ssl.client-auth=required` or `quarkus.http.ssl.client-auth=request`.
+     * <p/>
+     * Properties file is expected to have the `CN=role1,role,...,roleN` format.
```
Pursuant to my subsequent comment:

```suggestion
     * Properties file is expected to have the `CN=role1,role,...,roleN` format and should be encoded using UTF-8.
```

dmlloyd commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:
```diff
@@ -362,4 +373,50 @@ private synchronized void setPatchMatchingPolicyEnabled() {
+        for (String s : value.split(",")) {
+            roles.add(s.trim());
+        }
+        return roles;
```
It might be a good idea to either do this stream-wise or return `Set.copyOf(roles)` just so that the long-lived set is as small as possible. Obviously this would only make a difference if there were really a lot of roles though, so it's up to you.

dmlloyd reviewed (commented):
LGTM, for what that's worth, except for one problem around loading the properties file.

(In case anyone is wondering: I reviewed this code mainly because I was curious having tinkered with this kind of thing a bit back in my days contributing to the WildFly Elytron project.)

sberyozkin commented:
@dmlloyd Hi David, I think I've addressed your comments, which was easy since your typed how to address them :-), if there is anything else that you'd like to change, let me know please

sberyozkin commented:
Thanks @dmlloyd @michalvavrik and @Ladicek for the reviews. Let me merge and as always, things can be improved/optimized going forward

cescoffier commented:
@sberyozkin Just made aware of this. Why didn't you reuse the mechanism defined in https://datatracker.ietf.org/doc/html/rfc5755#section-4.4.5 instead of using a custom attribute (which lead to interoperability issue)

cescoffier commented:
At least it should be `X-Roles` (or a valid syntax).

sberyozkin commented:
@cescoffier Sorry, I have missed it.

>Why didn't you reuse the mechanism defined in https://datatracker.ietf.org/doc/html/rfc5755#section-4.4.5 instead of using a custom attribute (which lead to interoperability issue)

We don't expect here `CName` to contain any roles, the requirement this PR addresses is only about using Cname as a key to pick up the deployment specific roles from the local file.

As far as expecting the certificate already have the roles, in the `X-Roles` attribute, please open an enhancement request if you'd like, interesting idea, it won't be in conflict with this PR

sberyozkin commented:
@cescoffier This PR addresses a specific requirement, use `CName` as a map key.

But sure, we can make the attribute name configurable, sounds good to me.
